### PR TITLE
feat: timeline v2

### LINKED ISSUE #799: Combined feed items
Currently there is a lot of repetition on the feed - every time someone you follow comments, collects, mirrors or likes a post, it creates a duplicate item on your feed.

Recent API additions allow us to bundle these into a single item, here's the proposed design:

**Combined comments from people you follow** 

Show the latest one at the bottom, and collapse the rest into "Username and X others commented

Image

**Combined mirrors/collects/reactions from people you follow**

Image

Image

Image

When there are multiple combined actions, we can combine them so it doesn't get too verbose:

Image

And these actions can also be combined with multiple comments in the same feed item:

Image

### PR DESCRIPTION
**Pull request by sasicodes**
## What does this PR do?

Fixes #799 

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Enhancement (non-breaking small changes to existing functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How should this be tested?

- [ ] Test A
- [ ] Test B

### CODE REVIEW

File: src/components/Home/index.tsx [hunk 1]
```diff
@@ -35,7 +36,7 @@ const Home: NextPage = () => {
           {currentProfile ? (
             <>
               <NewPost />
-              <HomeFeed />
+              {isFeatureEnabled('timeline-v2', currentProfile?.id) ? <V2Timeline /> : <HomeFeed />}
```
Comment (bigint) [inline src/components/Home/index.tsx:39]: **suggestion:** better rename to `TimelineV2`

File: src/graphql/queries/V2HomeFeed.graphql [hunk 2]
```diff
@@ -0,0 +1,55 @@
+query V2HomeFeed(
```
Comment (bigint) [inline src/graphql/queries/V2HomeFeed.graphql:1]: **suggestion:** better we can use timeline instead of home feed everywhere

**Review by bigint — CHANGES_REQUESTED** [review]

**Review by bigint — CHANGES_REQUESTED** [review]

File: src/components/Timeline/Publication/Actions/Collect/CollectModule.tsx [hunk 3]
```diff
@@ -0,0 +1,422 @@
+import { LensHubProxy } from '@abis/LensHubProxy';
```
Comment (bigint) [inline src/components/Timeline/Publication/Actions/Collect/CollectModule.tsx:1]: **question:** can we reuse this from `Publication` folder?

File: src/components/Timeline/Publication/Actions/Collect/ReferralAlert.tsx [hunk 4]
```diff
@@ -0,0 +1,28 @@
+import Slug from '@components/Shared/Slug';
```
Comment (bigint) [inline src/components/Timeline/Publication/Actions/Collect/ReferralAlert.tsx:1]: **question:** can we reuse this from `Publication` folder?

File: src/components/Timeline/Publication/Actions/Collect/index.tsx [hunk 5]
```diff
@@ -0,0 +1,98 @@
+import Loader from '@components/Shared/Loader';
```
Comment (bigint) [inline src/components/Timeline/Publication/Actions/Collect/index.tsx:1]: **question:** can we reuse this from `Publication` folder?

File: src/components/Timeline/Publication/Actions/index.tsx [hunk 6]
```diff
@@ -0,0 +1,50 @@
+import Comment from '@components/Publication/Actions/Comment';
```
Comment (bigint) [inline src/components/Timeline/Publication/Actions/index.tsx:1]: **question:** can we reuse this from `Publication` folder?

We can make all those changes there to support `electedMirror` as optional props

File: src/components/Timeline/Publication/SinglePublication.tsx [hunk 7]
```diff
@@ -0,0 +1,78 @@
+import ModAction from '@components/Publication/Actions/ModAction';
```
Comment (bigint) [inline src/components/Timeline/Publication/SinglePublication.tsx:1]: **question:** can we reuse this from `Publication` folder?

File: src/components/Publication/Actions/Comment.tsx [hunk 8]
```diff
@@ -14,14 +14,14 @@ interface Props {
...
-    publication.__typename === 'Mirror'
+    publication?.__typename === 'Mirror'
       ? publication?.mirrorOf?.stats?.totalAmountOfComments
       : publication?.stats?.totalAmountOfComments;
   const iconClassName = isFullPublication ? 'w-[17px] sm:w-[20px]' : 'w-[15px] sm:w-[18px]';
 
   return (
     <motion.button whileTap={{ scale: 0.9 }} aria-label="Comment">
-      <Link href={`/posts/${publication.id}`}>
+      <Link href={`/posts/${publication?.id}`}>
```
Comment (bigint) [inline src/components/Publication/Actions/Comment.tsx:24]: **question:** why optional chain here? `id` will be always there 🤔

File: src/components/Publication/Actions/Comment.tsx [hunk 9]
```diff
@@ -14,14 +14,14 @@ interface Props {
 
 const Comment: FC<Props> = ({ publication, isFullPublication }) => {
   const count =
-    publication.__typename === 'Mirror'
+    publication?.__typename === 'Mirror'
```
Comment (bigint) [inline src/components/Publication/Actions/Comment.tsx:17]: **nitpick:** `__typename` will be always there 🤔  cuz apollo injects it

File: src/components/Publication/Actions/Like.tsx [hunk 10]
```diff
@@ -24,7 +24,7 @@ interface Props {
 }
 
 const Like: FC<Props> = ({ publication, isFullPublication }) => {
-  const isMirror = publication.__typename === 'Mirror';
+  const isMirror = publication?.__typename === 'Mirror';
```
Comment (bigint) [inline src/components/Publication/Actions/Like.tsx:27]: **nitpick:** `__typename` will be always there 🤔  cuz apollo injects it

File: src/components/Publication/Actions/Mirror.tsx [hunk 11]
```diff
@@ -31,7 +31,7 @@ interface Props {
 }
 
 const Mirror: FC<Props> = ({ publication, isFullPublication }) => {
-  const isMirror = publication.__typename === 'Mirror';
+  const isMirror = publication?.__typename === 'Mirror';
```
Comment (bigint) [inline src/components/Publication/Actions/Mirror.tsx:34]: **nitpick:** `__typename` will be always there 🤔  cuz apollo injects it

**Review by bigint — CHANGES_REQUESTED** [review]
@sasicodes left some comments 🙇🏼

**Comment by bigint** [comment]
Screenshot 2022-10-14 at 10 14 48 PM

We can remove space before comma

`Sasi , blah and blah` => `Sasi, blah and blah`

File: src/components/Timeline/Publication/Actions/index.tsx (hunk 6 above)
Comment (sasicodes) [reply src/components/Timeline/Publication/Actions/index.tsx]: Done

File: src/components/Timeline/Publication/Actions/Collect/index.tsx (hunk 5 above)
Comment (sasicodes) [reply src/components/Timeline/Publication/Actions/Collect/index.tsx]: Done

File: src/components/Timeline/Publication/Actions/Collect/ReferralAlert.tsx (hunk 4 above)
Comment (sasicodes) [reply src/components/Timeline/Publication/Actions/Collect/ReferralAlert.tsx]: Done

File: src/components/Timeline/Publication/Actions/Collect/CollectModule.tsx (hunk 3 above)
Comment (sasicodes) [reply src/components/Timeline/Publication/Actions/Collect/CollectModule.tsx]: Done

File: src/components/Home/FeedEventFilters.tsx [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+  const feedEventFilters = useTimelinePersistStore((state) => state.feedEventFilters);
+  const setFeedEventFilters = useTimelinePersistStore((state) => state.setFeedEventFilters);
+
+  const handleChange = (e: ChangeEvent<HTMLInputElement>) => {
+    setFeedEventFilters({ ...feedEventFilters, [e.target.name]: e.target.checked });
+  };
+
+  return (
+    <div className="flex justify-end">
+      <div className="flex items-center space-x-4">
```
Comment (bigint) [inline src/components/Home/FeedEventFilters.tsx:21]: **question:** why flex for single children?

File: src/components/Home/FeedType.tsx [hunk 13]
```diff
@@ -1,37 +1,46 @@
...
 import { Leafwatch } from '@lib/leafwatch';
 import type { Dispatch, FC } from 'react';
+import { useAppStore } from 'src/store/app';
 import { MISCELLANEOUS } from 'src/tracking';
 
+import FeedEventFilters from './FeedEventFilters';
+
 interface Props {
   setFeedType: Dispatch<'TIMELINE' | 'HIGHLIGHTS'>;
   feedType: 'TIMELINE' | 'HIGHLIGHTS';
```
Comment (bigint) [inline src/components/Home/FeedType.tsx:13]: **suggestion:** better move to lenstertypes.ts as enum so we can prevent using magic string inside jsx also using magic string is very bad 🙏🏼

File: src/components/Home/FeedType.tsx [hunk 14]
```diff
@@ -1,37 +1,46 @@
...
+          icon={<SparklesIcon className="w-4 h-4" />}
+          active={feedType === 'HIGHLIGHTS'}
+          showOnSm
+          onClick={() => {
+            setFeedType('HIGHLIGHTS');
+            Leafwatch.track(MISCELLANEOUS.SWITCH_HIGHLIGHTS);
+          }}
+        />
+      </div>
+      {feedType === 'TIMELINE' && isFeatureEnabled('timeline-v2', currentProfile?.id) && <FeedEventFilters />}
```
Comment (bigint) [inline src/components/Home/FeedType.tsx:43]: **nitpick:** no magic string bls 🙏🏼

File: src/components/Home/FeedType.tsx [hunk 15]
```diff
@@ -1,37 +1,46 @@
...
+            Leafwatch.track(MISCELLANEOUS.SWITCH_TIMELINE);
+          }}
+        />
+        <TabButton
+          name="Highlights"
+          icon={<SparklesIcon className="w-4 h-4" />}
+          active={feedType === 'HIGHLIGHTS'}
+          showOnSm
+          onClick={() => {
+            setFeedType('HIGHLIGHTS');
```
Comment (bigint) [inline src/components/Home/FeedType.tsx:38]: **nitpick:** no magic string bls 🙏🏼

File: src/components/Home/index.tsx [hunk 16]
```diff
@@ -40,7 +41,15 @@ const Home: NextPage = () => {
             <>
               <NewPost />
               <FeedType feedType={feedType} setFeedType={setFeedType} />
-              {feedType === 'TIMELINE' ? <HomeFeed /> : <Highlights />}
+              {feedType === 'TIMELINE' ? (
```
Comment (bigint) [inline src/components/Home/index.tsx:44]: **nitpick:** no magic string bls 🙏🏼

File: src/components/Publication/ThreadBody.tsx [hunk 17]
```diff
@@ -45,7 +45,7 @@ const ThreadBody: FC<Props> = ({ publication }) => {
           ) : (
             <>
               <PublicationBody publication={publication} />
-              <PublicationActions publication={publication} />
+              {publication && <PublicationActions publication={publication} />}
```
Comment (bigint) [inline src/components/Publication/ThreadBody.tsx:48]: **question:** publication will be always there right?

File: src/components/Publication/ThreadBody.tsx [hunk 18]
```diff
@@ -15,11 +15,11 @@ dayjs.extend(relativeTime);
 
 interface Props {
   publication: LensterPublication;
+  isMirror?: boolean;
```
Comment (bigint) [inline src/components/Publication/ThreadBody.tsx:18]: **question:** why we need isMirror check?

File: src/store/timeline.ts [hunk 19]
```diff
@@ -0,0 +1,17 @@
+import create from 'zustand';
+import { persist } from 'zustand/middleware';
+
+interface TransactionPersistState {
```
Comment (bigint) [inline src/store/timeline.ts:4]: **nitpick:** lol change the typename

**Review by bigint — CHANGES_REQUESTED** [review]

File: src/components/Home/FeedType.tsx (hunk 13 above)
Comment (sasicodes) [reply src/components/Home/FeedType.tsx:13]: dei, idhu un code than

File: src/components/Timeline/Publication/EventType/index.tsx [hunk 20]
```diff
@@ -0,0 +1,76 @@
...
+  //     {feedItem.mirrors.length && !isComment ? <Mirrored mirrors={feedItem.mirrors} /> : null}
+  //     {(isComment || commentsCount > 0) && showThread && <Commented feedItem={feedItem} />}
+  //     {feedItem.collects.length && !isComment ? (
+  //       <Collected isComment={isComment} collects={feedItem.collects} />
+  //     ) : null}
+  //     {feedItem.reactions.length && !isComment ? (
+  //       <Reacted isComment={isComment} reactions={feedItem.reactions} />
+  //     ) : null}
+  //   </>
+  // );
```
[the remark this answers is not in the document]
Comment (sasicodes) [reply src/components/Timeline/Publication/EventType/index.tsx:54]: ayo, maranta. removing

File: src/components/Home/FeedEventFilters.tsx (hunk 12 above)
Comment (sasicodes) [reply src/components/Home/FeedEventFilters.tsx]: clear clear

**Comment by bigint** [comment]
Clicking the thread opening different post

https://user-images.githubusercontent.com/69431456/197706871-358cdb6d-4cd6-4826-b841-e902c80658ce.mov

**Comment by bigint** [comment]
Screenshot 2022-10-25 at 12 43 41 PM

Only icons are not explanatory 😕 

cc @jounih

File: src/components/UI/InputCheckbox.tsx [hunk 21]
```diff
@@ -0,0 +1,33 @@
+import clsx from 'clsx';
+import type { ComponentProps } from 'react';
+import { forwardRef, useId } from 'react';
+
+interface Props extends Omit<ComponentProps<'input'>, 'prefix'> {
+  label?: string;
+  className?: string;
+}
+
+export const InputCheckbox = forwardRef<HTMLInputElement, Props>(function CheckBox(
```
Comment (bigint) [inline src/components/UI/InputCheckbox.tsx:10]: **suggestion:** just name it checkbox

**Comment by sasicodes** [comment]
> Clicking the thread opening different post
> 
>  Screen_Recording_2022-10-25_at_12.39.59_PM.mov

Fixed

File: src/components/Timeline/index.tsx [hunk 22]
```diff
@@ -0,0 +1,108 @@
...
+    if (feedEventFilters.collects) {
+      filters.push(...[FeedEventItemType.CollectPost, FeedEventItemType.CollectComment]);
+    }
+    if (feedEventFilters.mirrors) {
+      filters.push(FeedEventItemType.Mirror);
+    }
+    if (feedEventFilters.reactions) {
+      filters.push(...[FeedEventItemType.ReactionPost, FeedEventItemType.ReactionComment]);
+    }
+    return filters;
```
Comment (bigint) [inline src/components/Timeline/index.tsx:40]: better we can use switch right?

File: src/components/Timeline/index.tsx [hunk 23]
```diff
@@ -0,0 +1,108 @@
...
+              <div key={txn.id}>
+                <QueuedPublication txn={txn} />
+              </div>
+            )
+        )}
+        {publications?.map((publication, index: number) => (
+          <SinglePublication
+            feedItem={publication as FeedItem}
+            publication={publication.root as LensterPublication}
+            key={`${publication?.root.id}_${index}`}
```
Comment (bigint) [inline src/components/Timeline/index.tsx:100]: make key as first prop

**Comment by jounih** [comment]
> Screenshot 2022-10-25 at 12 43 41 PM
> 
> Screenshot 2022-10-25 at 12 44 15 PM
> 
> Only icons are not explanatory 😕
> 
> cc @jounih

@bigint @sasicodes  we could go back to tex - "Commented and liked by" etc - it's longer but maybe easier to understand

File: src/components/Timeline/Publication/EventType/ProfileCircles.tsx [hunk 24]
```diff
@@ -0,0 +1,86 @@
+import type { Profile } from '@generated/types';
+import getAvatar from '@lib/getAvatar';
+import Link from 'next/link';
+import type { FC, ReactNode } from 'react';
+
+interface Props {
+  profiles: Profile[];
+  totalCount: number;
```
Comment (bigint) [inline src/components/Timeline/Publication/EventType/ProfileCircles.tsx:8]: do we need this prop?

File: src/components/Publication/SinglePublication.tsx [hunk 25]
```diff
@@ -25,19 +28,34 @@ interface Props {
...
+    ? rootPublication.createdAt
+    : isMirror
+    ? publication?.mirrorOf?.createdAt
+    : publication?.createdAt;
 
   return (
     <article className="hover:bg-gray-100 dark:hover:bg-gray-800 cursor-pointer first:rounded-t-xl last:rounded-b-xl p-5">
-      <PublicationType publication={publication} showType={showType} showThread={showThread} />
+      {feedItem ? (
+        <EventType feedItem={feedItem} showType={showType} showThread={showThread} />
```
Comment (bigint) [inline src/components/Publication/SinglePublication.tsx:55]: `isComment` prop is not used anywhere, better remove in all places

Comment (bigint) [reply src/components/Publication/SinglePublication.tsx:55]: chorry

File: src/components/Timeline/EventType/Commented.tsx [hunk 26]
```diff
@@ -0,0 +1,25 @@
...
+interface Props {
+  feedItem: FeedItem;
+}
+
+const Commented: FC<Props> = ({ feedItem }) => {
+  const publication = feedItem.root as Comment;
+  const firstComment = feedItem.comments && feedItem.comments[0];
+
+  return (
+    <span>
```
Comment (bigint) [inline src/components/Timeline/EventType/Commented.tsx:15]: do we need span here?

File: src/components/Timeline/EventType/Collected.tsx [hunk 27]
```diff
@@ -0,0 +1,34 @@
...
+    profiles = profiles.filter(
+      (profile, index, self) => index === self.findIndex((t) => t.id === profile.id)
+    );
+    return profiles;
+  };
+
+  return (
+    <div
+      className={clsx('flex items-center pb-4 space-x-1 text-gray-500 text-[13px]', {
+        'ml-[45px] !pb-2': isComment
```
Comment (bigint) [inline src/components/Timeline/EventType/Collected.tsx:25]: will this be used anywhere? wanna see the screenshot 🙇

File: src/components/Timeline/EventType/Mirrored.tsx [hunk 28]
```diff
@@ -0,0 +1,34 @@
...
+    profiles = profiles.filter(
+      (profile, index, self) => index === self.findIndex((t) => t.id === profile.id)
+    );
+    return profiles;
+  };
+
+  return (
+    <div
+      className={clsx('flex items-center pb-4 space-x-1 text-gray-500 text-[13px]', {
+        'ml-[45px] !pb-2': isComment
```
Comment (bigint) [inline src/components/Timeline/EventType/Mirrored.tsx:25]: this one too

File: src/components/Timeline/EventType/Reacted.tsx [hunk 29]
```diff
@@ -0,0 +1,34 @@
...
+    profiles = profiles.filter(
+      (profile, index, self) => index === self.findIndex((t) => t.id === profile.id)
+    );
+    return profiles;
+  };
+
+  return (
+    <div
+      className={clsx('flex items-center pb-4 space-x-1 text-gray-500 text-[13px]', {
+        'ml-[45px] !pb-2': isComment
```
Comment (bigint) [inline src/components/Timeline/EventType/Reacted.tsx:25]: this one too

**Review by bigint — APPROVED** [review]
**Code LGTM** :shipit: 

Rad LGTM

**Pull request merged**